### PR TITLE
chore: upgrade example to RN 0.74, move to typescript, update deprecated code

### PR DESCRIPTION
**Pull request by okwasniewski**
### Summary

- Upgrade example app to RN 0.74
- Update the deprecated method call for `[[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:_entryFile];` as the `fallbackResource` is not accepted anymore
- Use react-native-builder-bob
- Use release-it
- Migrate to typescript from Flow
- Adjust Android source code to work on RN 0.74

### Test plan

Run all 4 example apps.

### CODE REVIEW

**Comment by okwasniewski** [comment]
The CI is failing due to too old node version (currently at node 10.24.1)

**Comment by atomicpages** [comment]
This is excellent 🙏 , thanks for putting this together.

**Comment by advyth** [comment]
Hi guys, any update on this PR. It would be great if we can get it merged.

File: package.json [hunk 1]
```diff
@@ -8,37 +8,140 @@
...
+    "!**/__tests__",
+    "!**/__fixtures__",
+    "!**/__mocks__",
+    "!**/.*"
+  ],
   "publishConfig": {
     "access": "public"
   },
+  "resolutions": {
+    "@types/react": "17.0.21"
```
Comment (mdjastrzebski) [inline package.json]: Is there any particular reason why this is not `v18`?

File: package.json [hunk 2]
```diff
@@ -8,37 +8,140 @@
...
+    "prettier": "^3.0.2",
+    "react": "18.2.0",
+    "react-native": "0.74.2",
+    "react-native-gesture-handler": "2.16.2",
+    "react-native-safe-area-context": "^4.10.4",
+    "react-native-screens": "^3.31.1",
+    "react-test-renderer": "18.2.0",
+    "release-it": "^15.0.0",
+    "@release-it/conventional-changelog": "^5.0.0",
+    "typescript": "5.0.4"
```
Comment (mdjastrzebski) [inline package.json]: [nit] Consider using latest, consider using `^` version

File: example/swift/.xcode.env.local [hunk 3]
```diff
@@ -0,0 +1,2 @@
+export NODE_BINARY=/opt/homebrew/bin/node
```
Comment (mdjastrzebski) [inline example/swift/.xcode.env.local]: Should this file be committed to the repo?

File: example/objc/PrivacyInfo.xcprivacy [hunk 4]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
```
Comment (mdjastrzebski) [inline example/objc/PrivacyInfo.xcprivacy:4]: Q: Is any of these actually required by RN Brownfield? If so, we should try to remove their usage in the future.

File: example/objc/.xcode.env.local [hunk 5]
```diff
@@ -0,0 +1,2 @@
+export NODE_BINARY=/opt/homebrew/bin/node
```
Comment (mdjastrzebski) [inline example/objc/.xcode.env.local]: Should this be committed to the repo?

File: example/kotlin/settings.gradle [hunk 6]
```diff
@@ -1,4 +1,5 @@
-apply from: file("../../node_modules/@react-native-community/cli-platform-android/native_modules.gradle");
-applyNativeModulesSettingsGradle(settings, "../..")
+rootProject.name = 'BrownfieldExample'
+apply from: file("../../node_modules/@react-native-community/cli-platform-android/native_modules.gradle"); applyNativeModulesSettingsGradle(settings)
```
Comment (mdjastrzebski) [inline example/kotlin/settings.gradle:2]: [nit]
```suggestion
apply from: file("../../node_modules/@react-native-community/cli-platform-android/native_modules.gradle")
applyNativeModulesSettingsGradle(settings)
```

File: example/kotlin/gradle.properties [hunk 7]
```diff
@@ -16,3 +16,18 @@ kotlin.code.style=official
...
+# You can also override it from the CLI using
+# ./gradlew <task> -PreactNativeArchitectures=x86_64
+reactNativeArchitectures=armeabi-v7a,arm64-v8a,x86,x86_64
+
+# Use this property to enable support to the new architecture.
+# This will allow you to use TurboModules and the Fabric render in
+# your application. You should enable this flag either if you want
+# to write custom TurboModules/Fabric components OR use libraries that
+# are providing them.
+newArchEnabled=false
```
Comment (mdjastrzebski) [inline example/kotlin/gradle.properties:29]: q: does RN Brownfield work now with New Arch?

File: example/kotlin/build.gradle [hunk 8]
```diff
@@ -2,44 +2,32 @@
...
+        targetSdkVersion = 34
+        ndkVersion = "26.1.10909125"
+        kotlinVersion = "1.9.22"
+
+        // We use NDK 23 which has both M1 support and is the side-by-side NDK version from AGP.
+        ndkVersion = "23.1.7779620"
     }
     repositories {
         google()
         jcenter()
```
Comment (mdjastrzebski) [inline example/kotlin/build.gradle]: [JCenter is deprecated](https://blog.gradle.org/jcenter-shutdown), can we remove this repo or is there still some dep that needs it?

File: example/kotlin/app/build.gradle [hunk 9]
```diff
@@ -30,19 +34,24 @@ android {
...
+    if (hermesEnabled.toBoolean()) {
+        implementation("com.facebook.react:hermes-android")
+    } else {
+        implementation jscFlavor
+    }
 }
 
-apply from: file("../../../node_modules/@react-native-community/cli-platform-android/native_modules.gradle");
-applyNativeModulesAppBuildGradle(project, "../..")
+apply from: file("../../../node_modules/@react-native-community/cli-platform-android/native_modules.gradle"); applyNativeModulesAppBuildGradle(project)
```
Comment (mdjastrzebski) [inline example/kotlin/app/build.gradle:57]: [nit]
```suggestion
apply from: file("../../../node_modules/@react-native-community/cli-platform-android/native_modules.gradle")
applyNativeModulesAppBuildGradle(project)
```

**Review by mdjastrzebski — APPROVED** [review]
LGTM, left some comments, but nothing major. Really happy to see that PR getting merged and RN Brownfield to be back 🙌

File: example/kotlin/settings.gradle (hunk 6 above)
Comment (mdjastrzebski) [reply example/kotlin/settings.gradle:2]: Is this some template generated `gradle.settings`? If so, please ignore the comment.

File: example/swift/.xcode.env.local (hunk 3 above)
Comment (jbroma) [reply example/swift/.xcode.env.local]: local should stay ignored, this is setup automatically during `pod install`

File: example/kotlin/settings.gradle (hunk 6 above)
Comment (okwasniewski) [reply example/kotlin/settings.gradle:2]: Yup, it's like this in the template.

File: example/kotlin/gradle.properties (hunk 7 above)
Comment (okwasniewski) [reply example/kotlin/gradle.properties:29]: Nope 👎 My Next PR will bring new arch support

File: example/objc/PrivacyInfo.xcprivacy (hunk 4 above)
Comment (okwasniewski) [reply example/objc/PrivacyInfo.xcprivacy:4]: Not directly, but this file is required by Apple now, so I think it's a good example to keep it. It was also generated during pod install

File: example/kotlin/app/build.gradle (hunk 9 above)
Comment (okwasniewski) [reply example/kotlin/app/build.gradle:57]: Same as above, this file is from generated template

File: example/objc/PrivacyInfo.xcprivacy (hunk 4 above)
Comment (mdjastrzebski) [reply example/objc/PrivacyInfo.xcprivacy:4]: So the requirements for that come from RN directly?

Comment (okwasniewski) [reply example/objc/PrivacyInfo.xcprivacy:4]: So requirement for this file comes directly from Apple, to publish your app you need to specify the Privacy Manifest. React Native generates this file with some cocoapods script. I don't want to add it to the .gitignore as this file should be committed. 

Here is a reference: https://medium.com/@dabhir16/a-guide-to-apples-privacy-manifests-in-2024-c3f476b3fa6f

**Review by jbroma — APPROVED** [review]
LGTM 🎉

PR is quite big so might find so things to fix after merging this

**Pull request merged**